File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public int dfsHeight(TreeNode root, int[] diameter)
    {
        if(root == null) return 0;
        
        int lh = dfsHeight(root.left, diameter);
        int rh = dfsHeight(root.right, diameter);
    
        diameter[0] = Math.max(diameter[0], lh + rh);
        
        return 1 + Math.max(lh, rh);
    }
    
    public int diameterOfBinaryTree(TreeNode root) {
        
        int[] diameter = new int[1];
        dfsHeight(root, diameter);
        return diameter[0];
    }
}

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int dfs(TreeNode* root, int& diameter) {
        
        if(root == NULL) return 0;
        
        int lh = dfs(root->left, diameter);
        int rh = dfs(root->right, diameter);
        
        diameter = max(lh+rh, diameter);
        
        return 1 + max(lh, rh);
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        
        int diameter = 0;
        dfs(root, diameter);
        return diameter;
    }
};